samza-core/src/main/java/org/apache/samza/util/EmbeddedTaggedRateLimiter.java

```diff
@@ -0,0 +1,130 @@
+    this.tagToTargetRateMap = tagToCreditsPerSecondMap;
+  }
+
+  @Override
+  public void acquire(Map<String, Integer> tagToCreditsMap) {
+    ensureTagsAreValid(tagToCreditsMap);
+    tagToCreditsMap.forEach((tag, numberOfCredits) -> tagToRateLimiterMap.get(tag).acquire(numberOfCredits));
+  }
+
+  @Override
+  public Map<String, Integer> acquire(Map<String, Integer> tagToCreditsMap, long timeout, TimeUnit unit) {
+    ensureTagsAreValid(tagToCreditsMap);
+
+    long timeoutInNanos = NANOSECONDS.convert(timeout, unit);
+
+    Stopwatch stopwatch = Stopwatch.createStarted();
+    return tagToCreditsMap.entrySet().stream()
+        .map(e -> {
+            String tag = e.getKey();
+            int requiredCredits = e.getValue();
+            long remainingTimeoutInNanos = Math.max(0L, timeoutInNanos - stopwatch.elapsed(NANOSECONDS));
+            com.google.common.util.concurrent.RateLimiter rateLimiter = tagToRateLimiterMap.get(tag);
+            int availableCredits = rateLimiter.tryAcquire(requiredCredits, remainingTimeoutInNanos, NANOSECONDS)
+                ? requiredCredits
+                : 0;
+            return new ImmutablePair<String, Integer>(tag, availableCredits);
+          })
+        .collect(Collectors.toMap(ImmutablePair::getKey, ImmutablePair::getValue));
+  }
+
+  @Override
+  public Map<String, Integer> tryAcquire(Map<String, Integer> tagToCreditsMap) {
+    ensureTagsAreValid(tagToCreditsMap);
+    return tagToCreditsMap.entrySet().stream()
+        .map(e -> {
+            String tag = e.getKey();
+            int requiredCredits = e.getValue();
+            int availableCredits = tagToRateLimiterMap.get(tag).tryAcquire(requiredCredits)
+                ? requiredCredits
+                : 0;
+            return new ImmutablePair<String, Integer>(tag, availableCredits);
+          })
+        .collect(Collectors.toMap(ImmutablePair::getKey, ImmutablePair::getValue));
+  }
+
+  @Override
+  public void acquire(int numberOfCredits) {
+    throw new IllegalArgumentException("This method is not applicable");
+  }
+
+  @Override
+  public int acquire(int numberOfCredit, long timeout, TimeUnit unit) {
+    throw new IllegalArgumentException("This method is not applicable");
+  }
+
+  @Override
+  public int tryAcquire(int numberOfCredit) {
+    throw new IllegalArgumentException("This method is not applicable");
+  }
+
+  @Override
+  public void init(Config config, TaskContext taskContext) {
+    int numberOfTasks = taskContext.getSamzaContainerContext().taskNames.size();
+    this.tagToRateLimiterMap = Collections.unmodifiableMap(tagToTargetRateMap.entrySet().stream()
+        .map(e -> {
+            String tag = e.getKey();
+            Integer ratePerTask = e.getValue() / numberOfTasks;
+            return new ImmutablePair<String, com.google.common.util.concurrent.RateLimiter>(
+                tag, com.google.common.util.concurrent.RateLimiter.create(ratePerTask));
```

Comment:
nit: should we log the effective rate here?

Reply:
Fixed